apps/btc_family/btc_txn.c

```diff
@@ -253,24 +236,118 @@ static bool fetch_transaction_meta(btc_query_t *query) {
   // TODO: check if malloc failed; report to host and exit
   send_response(BTC_SIGN_TXN_RESPONSE_META_ACCEPTED_TAG);
   return true;
 }
 
+static bool handle_verify_input_utxo(btc_query_t *query) {
+  // Validate input UTXOs for safety from attack. Ref:
+  // https://blog.trezor.io/details-of-firmware-updates-for-trezor-one-version-1-9-1-and-trezor-model-t-version-2-3-1-1eba8f60f2dd
+  for (int idx = 0; idx < g_txn_context->metadata.input_count; idx++) {
+    if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) &&
+        !check_which_request(query, BTC_SIGN_TXN_REQUEST_INPUT_TAG)) {
+      return false;
+    }
+
+    const btc_sign_txn_input_prev_txn_t *txn = &query->sign_txn.input.prev_txn;
+    if (!btc_verify_utxo(txn->bytes, txn->size, &g_txn_context->inputs[idx])) {
+      // input UTXO validation failed, terminate immediately
+      btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+
+    // duplicate the input UTXO details into g_txn_context
+    btc_txn_input_t *input = &g_txn_context->inputs[idx];
+    input->prev_output_index = query->sign_txn.input.prev_output_index;
+    input->address_index = query->sign_txn.input.address_index;
+    input->change_index = query->sign_txn.input.change_index;
+    input->value = query->sign_txn.input.value;
+    input->sequence = query->sign_txn.input.sequence;
+    input->script_pub_key.size = query->sign_txn.input.script_pub_key.size;
+    memcpy(input->prev_txn_hash, query->sign_txn.input.prev_txn_hash, 32);
+    memcpy(input->script_pub_key.bytes,
+           query->sign_txn.input.script_pub_key.bytes,
+           input->script_pub_key.size);
+
+    // send accepted response to indicate validation of input passed
+    send_response(BTC_SIGN_TXN_RESPONSE_INPUT_ACCEPTED_TAG);
+  }
+  return true;
+}
+
+static bool handle_user_verification(btc_query_t *query) {
+  char title[20] = "";
+  char value[100] = "";
+  char address[100] = "";
+
+  for (int idx = 0; idx < g_txn_context->metadata.output_count; idx++) {
+    if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) &&
+        !check_which_request(query, BTC_SIGN_TXN_REQUEST_OUTPUT_TAG)) {
+      return false;
+    }
+
+    snprintf(title, sizeof(title), "Receiver #%d", (idx + 1));
+    btc_sign_txn_output_t *output = &query->sign_txn.output;
+    btc_sign_txn_output_script_pub_key_t *script = &output->script_pub_key;
+    memcpy(&g_txn_context->outputs[idx], output, sizeof(btc_sign_txn_output_t));
+    if (true == output->is_change) {
+      // do not show the change output UTXOs
+      continue;
+    }
+    format_value(output->value, value, sizeof(value));
+    script_output_to_address(
+        script->bytes, script->size, address, sizeof(address));
+    if (!core_scroll_page(title, address, btc_send_error) ||
+        !core_scroll_page(title, value, btc_send_error)) {
+      return false;
+    }
```

Comment:
Is possible, let's to keep user verification independent of the data flow with the usb host

Reply:
split into 2